docs/utxo-hd/v0.1/for-developers.md

```diff
@@ -0,0 +1,456 @@
+     LedgerCfg l
+  -> SlotNo
+  -> l EmptyMK
+  -> LedgerResult l (Ticked1 l DiffMK)
+```
+
+### Ticking and era boundaries
+
+In general, ticking a ledger state won't modify the values in the UTxO set, but
+there are two exceptions in the Cardano case when ticking crosses an era
+boundary: 
+
+- For simplicity, we have deliberately chosen that the Byron era would not have
+  any ledger tables, and there would only be ledger tables starting in Shelley.
+  This means that when crossing the boundary to Shelley, the whole UTxO set will
+  be returned as deltas (see `translateLedgerStateByronToShelleyWrapper`
+  [here](../../../ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/CanHardFork.hs)).
+
+- In the Shelley to Allegra transition, a bunch of UTxOs (owned by the AVVM
+  addresses) were consumed into the treasury. This transition will return said
+  UTxOs as deletions (see `translateLedgerStateShelleyToAllegraWrapper`
+  [here](../../../ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/CanHardFork.hs)).
+
+The general scheme for **ticking a block is: a ledger state without a
+UTxO set (`EmptyMK`) goes in and a ledger state with some deltas (`DiffMK`)
+comes out**. This is reflected in `applyChainTickLedgerResult`.
+
+```haskell
+applyChainTickLedgerResult ::
+     LedgerCfg l
+  -> SlotNo
+  -> l EmptyMK
+  -> LedgerResult l (Ticked1 l DiffMK)
+```
+
+As one might have noticed, the diagram for applying a block is missing the
+ticking step, but it would complicate the diagram too much. The actual flow is:
+
+- The `emptyInputState` is ticked, producing a `Ticked1 l DiffMK`.
+- The `forwardedTables` are also forwarded through the diffs coming from
+  ticking.
+- The diffs from ticking are prepended to the resulting diffs before being
+  stored in the `DbChangelog`.
+
+### The mempool
+
+The mempool behaves pretty much as a virtual block. The design is not
+particularly complex as we just ask the backing store for the inputs for a
+transaction when applying it. The only caveat compared to the pre-UTxO-HD
+implementation is that previously it cached the latest ledger state and
+therefore we could run a separate thread that would sync the mempool with the
+LedgerDB and revalidate all the transactions when necessary, but now we might
+not be able to apply a transaction if the ledger state on top of which we had
+applied the others is now gone from the LedgerDB. Therefore, adding a
+transaction might in some cases trigger a sync with the LedgerDB and therefore a
+revalidation of the previous transactions.
+
+### Ledger state queries
+
+Most of the queries don't require the UTxO set but there are three in particular
+that do: `GetUTxOByTxIn`, `GetUTxOWhole` and `GetUTxOByAddress`. We assume that
+`GetUTxOWhole` is a debug query so we don't pay much attention to it as we
+expect it will be used with small-ish UTxO sets. For `GetUTxOByTxIn`, the story
```

Comment:
Even if `GetUTxOWhole` is used with a large set, that's the user's problem, not ours, right? They chose to use a `Debug` query.

So I don't know that the mention of small UTxO sets is helpful here.

Reply:
Yeah, good point.